content/blog/2019-11-14-mempass/index.md

```diff
@@ -0,0 +1,292 @@
+
+On the other hand, our second call to `free` occurs attempts to free `%12`, which is the same address as `%10`. MemPass simply checks the allocation table, and since the address is no longer stored here we have a double free bug. MemPass prints this as a warning to the console, and continues searching for vulnerabilities.
+
+While MemPass doesn't handle calls to `realloc`, it would be quite tractible to do so now. Just remove the old memory address from the allocation hash, and add the new address that the function returns (along with its size).
+
+## Use After Free: Accessing invalid memory
+
+One of the more difficult aspects of MemPass's implementation is to find a way to handle accesses to memory after a pointer has been freed. Consider the following program:
+
+```c
+int* ptr;
+ptr = (int *)malloc(4 * sizeof(int));
+free(ptr);
+ptr[0] = 4;
+return 0;
+```
+
+The naïve solution would be to add instrumentation after every load or store instruction in the LLVM IR, and compare the addresses to our allocation table. However, we run into a series of complications once we look at the actual IR.
+
+```
+%3 = alloca i32, align 4
+%4 = alloca i32, align 4
+%5 = alloca i8**, align 8
+%6 = alloca i32*, align 8
+store i32 0, i32* %3, align 4
+store i32 %0, i32* %4, align 4
+store i8** %1, i8*** %5, align 8
+%7 = call noalias i8* @malloc(i64 16) #3
+%8 = bitcast i8* %7 to i32*
+store i32* %8, i32** %6, align 8
+%9 = load i32*, i32** %6, align 8
+%10 = bitcast i32* %9 to i8*
+call void @free(i8* %10) #3
+%11 = load i32*, i32** %6, align 8
+%12 = getelementptr inbounds i32, i32* %11, i64 0
+store i32 4, i32* %12, align 4
+ret i32 0
+```
+
+First, the program accesses more memory than what was allocated by the programmer through calls to `malloc`. If MemPass compares the address accessed by one of the first `store` instructions with the allocation hash, it wouldn't find the address and emit a false-positive warning.
+
+In reality, programs also allocates stack frame memory with `alloca` instructions. In order to handle these extra allocations, MemPass adds extra instrumentation here, and adds the stack frame addresses/sizes onto the allocation table.
+
+However, there's another problem. Some of the load and store instructions use pointers of arbitrary types. If MemPass doesn't know what the pointer types are, it can't pass those addresses to the runtime library for evaluation.
+
+A solution that MemPass employs is to insert `bitcast` instructions after every load or store instruction, converting the address pointer from its arbitrary type to an 8-bit integer pointer. Since we're only comparing addresses and not the actual values at these addresses, this should work somewhat well. However, the breadth of types in LLVM could allow for some imprecision here.

```

Comment:
Hmm… I don't quite see what the imprecision is. If you supply the base address (as an `i8*`) and the access size, I think you should have everything you need to check safety? Maybe I'm missing something.